### PR TITLE
[Modules] Adding new module for Microsoft.Insights/dataCollectionRules

### LINKED ISSUE #2542: [Feature Request]: Add module for Microsoft.Insights/dataCollectionRules
### Description

Currently there is no module for data collection rules (Microsoft.Insights/dataCollectionRules), which are needed for the data collection process in Azure Monitor. The template reference can be found [here](https://learn.microsoft.com/en-us/azure/templates/microsoft.insights/datacollectionrules?pivots=deployment-language-bicep).

### PR DESCRIPTION
**Pull request by krbar**
# Description

This PR resolves #2542 and adds a new module for `Microsoft.Insights/dataCollectionRules`.

| Pipeline |
| - |
| [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=131&branchName=users%2Fkrbar%2F2542-dataCollectionRulesModule) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.Insights/dataCollectionRules/readme.md [hunk 1]
```diff
@@ -0,0 +1,1628 @@
+# Insights DataCollectionRules `[Microsoft.Insights/dataCollectionRules]`
+
+This module deploys Insights DataCollectionRules.
+// TODO: Replace Resource and fill in description
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/readme.md:4]:
```suggestion
This module deploys DataCollection Rules.
```

File: modules/Microsoft.Insights/dataCollectionRules/readme.md [hunk 2]
```diff
@@ -0,0 +1,1628 @@
...
+| `location` | string | `[resourceGroup().location]` |  | Location for all Resources. |
+| `lock` | string | `''` | `['', CanNotDelete, ReadOnly]` | Specify the type of lock. |
+| `roleAssignments` | array | `[]` |  | Array of role assignment objects that contain the 'roleDefinitionIdOrName' and 'principalId' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: '/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11'. |
+| `streamDeclarations` | object | `{object}` |  | Declaration of custom streams used in this rule. |
+| `tags` | object | `{object}` |  | Resource tags. |
+
+
+### Parameter Usage: `<ParameterPlaceholder>`
+
+// TODO: Fill in Parameter usage
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/readme.md:50]: Leftover

File: modules/Microsoft.Insights/dataCollectionRules/deploy.bicep [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  location: location
+  name: name
+  tags: tags
+  properties: {
+    dataSources: dataSources
+    destinations: destinations
+    dataFlows: dataFlows
+    dataCollectionEndpointId: empty(dataCollectionEndpointId) ? null : dataCollectionEndpointId
+    streamDeclarations: empty(streamDeclarations) ? null : streamDeclarations
+    description: empty(dataCollectionRuleDescription) ? null : dataCollectionRuleDescription
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/deploy.bicep:80]:
```suggestion
    dataCollectionEndpointId: !empty(dataCollectionEndpointId) ? dataCollectionEndpointId : null
    streamDeclarations: !empty(streamDeclarations) ? streamDeclarations : null
    description: !empty(dataCollectionRuleDescription) ? dataCollectionRuleDescription : null
```

File: modules/Microsoft.Insights/dataCollectionRules/.test/custombasic/dependencies.bicep [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+output logAnalyticsWorkspaceName string = logAnalyticsWorkspace.name
+
+@description('The principal ID of the created managed identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created Managed Identity.')
+output managedIdentityResourceId string = managedIdentity.id
+
+@description('The resource ID of the created Data Collection Endpoint.')
+output dataCollectionEndpointResourceId string = dataCollectionEndpoint.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/.test/custombasic/dependencies.bicep:66]:
```suggestion
@description('The principal ID of the created managed identity.')
output managedIdentityPrincipalId string = managedIdentity.properties.principalId

@description('The resource ID of the created Data Collection Endpoint.')
output dataCollectionEndpointResourceId string = dataCollectionEndpoint.id
```

Appears not to be used.

File: modules/Microsoft.Insights/dataCollectionRules/.test/customadv/dependencies.bicep [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+output logAnalyticsWorkspaceName string = logAnalyticsWorkspace.name
+
+@description('The principal ID of the created managed identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created Managed Identity.')
+output managedIdentityResourceId string = managedIdentity.id
+
+@description('The resource ID of the created Data Collection Endpoint.')
+output dataCollectionEndpointResourceId string = dataCollectionEndpoint.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/.test/customadv/dependencies.bicep:82]:
```suggestion
@description('The principal ID of the created managed identity.')
output managedIdentityPrincipalId string = managedIdentity.properties.principalId

@description('The resource ID of the created Data Collection Endpoint.')
output dataCollectionEndpointResourceId string = dataCollectionEndpoint.id
```

Seems not to be used

File: modules/Microsoft.Insights/dataCollectionRules/.test/customiis/dependencies.bicep [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+output logAnalyticsWorkspaceName string = logAnalyticsWorkspace.name
+
+@description('The principal ID of the created managed identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created Managed Identity.')
+output managedIdentityResourceId string = managedIdentity.id
+
+@description('The resource ID of the created Data Collection Endpoint.')
+output dataCollectionEndpointResourceId string = dataCollectionEndpoint.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/.test/customiis/dependencies.bicep:47]:
```suggestion
@description('The principal ID of the created managed identity.')
output managedIdentityPrincipalId string = managedIdentity.properties.principalId

@description('The resource ID of the created Data Collection Endpoint.')
output dataCollectionEndpointResourceId string = dataCollectionEndpoint.id
```
it appears it isn't used

File: modules/Microsoft.Insights/dataCollectionRules/.test/linux/dependencies.bicep [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+output logAnalyticsWorkspaceResourceId string = logAnalyticsWorkspace.id
+
+@description('The name of the deployed log analytics workspace.')
+output logAnalyticsWorkspaceName string = logAnalyticsWorkspace.name
+
+@description('The principal ID of the created managed identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created Managed Identity.')
+output managedIdentityResourceId string = managedIdentity.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/.test/linux/dependencies.bicep:30]:
```suggestion
@description('The principal ID of the created managed identity.')
output managedIdentityPrincipalId string = managedIdentity.properties.principalId
```

It appears it isn't used

File: modules/Microsoft.Insights/dataCollectionRules/.test/windows/dependencies.bicep [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+output logAnalyticsWorkspaceResourceId string = logAnalyticsWorkspace.id
+
+@description('The name of the deployed log analytics workspace.')
+output logAnalyticsWorkspaceName string = logAnalyticsWorkspace.name
+
+@description('The principal ID of the created managed identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created Managed Identity.')
+output managedIdentityResourceId string = managedIdentity.id
```
Comment (AlexanderSehr) [inline modules/Microsoft.Insights/dataCollectionRules/.test/windows/dependencies.bicep:30]:
```suggestion
@description('The principal ID of the created managed identity.')
output managedIdentityPrincipalId string = managedIdentity.properties.principalId
```

File: modules/Microsoft.Insights/dataCollectionRules/readme.md (hunk 2 above)
Comment (krbar) [reply modules/Microsoft.Insights/dataCollectionRules/readme.md]: Fixed

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**